server/shared/src/main/scala/org/http4s/server/middleware/ResponseLogger.scala

```diff
@@ -31,121 +32,165 @@ import cats.~>
 import fs2.Chunk
 import fs2.Pipe
 import fs2.Stream
+import org.http4s.server.middleware.ResponseLogger.Lift
 import org.log4s.getLogger
 import org.typelevel.ci.CIString
 
+sealed abstract class ResponseLogger[F[_]] {
+  def apply[G[_], A](fk: F ~> G)(
+      http: Kleisli[G, A, Response[F]]
+  )(implicit G: MonadCancelThrow[G]): Kleisli[G, A, Response[F]]
+
+  def apply[G[_], A](
+      http: Kleisli[G, A, Response[F]]
+  )(implicit lift: Lift[F, G], G: MonadCancelThrow[G]): Kleisli[G, A, Response[F]] =
+    apply(lift.fk)(http)
+
+  def withRedactHeadersWhen(f: CIString => Boolean): ResponseLogger[F]
+
+  def withLogAction(f: String => F[Unit]): ResponseLogger[F]
+  def withLogActionOpt(of: Option[String => F[Unit]]): ResponseLogger[F] =
+    of.fold(this)(withLogAction)
+}
+
 /** Simple middleware for logging responses as they are processed
   */
 object ResponseLogger {
-  private[this] val logger = getLogger
 
-  def apply[G[_], F[_], A](
+  private[middleware] final case class Impl[F[_]](
       logHeaders: Boolean,
-      logBody: Boolean,
-      fk: F ~> G,
+      logBodyText: Either[Boolean, Stream[F, Byte] => Option[F[String]]],
       redactHeadersWhen: CIString => Boolean = Headers.SensitiveHeaders.contains,
       logAction: Option[String => F[Unit]] = None,
-  )(
-      http: Kleisli[G, A, Response[F]]
-  )(implicit G: MonadCancelThrow[G], F: Async[F]): Kleisli[G, A, Response[F]] =
-    impl[G, F, A](logHeaders, Left(logBody), fk, redactHeadersWhen, logAction)(http)
+  )(implicit F: Async[F])
+      extends ResponseLogger[F] {
+    override def apply[G[_], A](fk: F ~> G)(
+        http: Kleisli[G, A, Response[F]]
+    )(implicit G: MonadCancelThrow[G]): Kleisli[G, A, Response[F]] = {
```

Comment:
The implementation is completely unchanged. In retrospect maybe I should have left it as a method on the object to reduce the diff.